### ISSUE TITLE
asyncssh Connection failure:Failed to read binary packet data!

### ISSUE THREAD

**Issue by easymavinmind** [ronf/asyncssh#393]
I tried to connect to sftp server with asyncssh (version 2.5.0), but it failed, giving me an error:

```
INFO:asyncssh:[conn=0] Beginning auth for user User1
DEBUG:asyncssh:[conn=0] Received authentication banner
DEBUG:asyncssh:[conn=0] Trying public key auth with rsa-sha2-256 key
DEBUG:asyncssh:[conn=0] Trying keyboard-interactive auth
INFO:asyncssh:[conn=0] Auth for user User1 succeeded
Connected <asyncssh.connection.SSHClientConnection object at 0x10236fe50>
DEBUG:asyncssh:[conn=0, chan=0] Set write buffer limits: low-water=16384, high-water=65536
INFO:asyncssh:[conn=0, chan=0] Requesting new SSH session
DEBUG:asyncssh:[conn=0] Received disconnect: Failed to read binary packet data! (2)
INFO:asyncssh:[conn=0] Connection failure: Failed to read binary packet data!
INFO:asyncssh:[conn=0, chan=0] Closing channel due to connection close
INFO:asyncssh:[conn=0, chan=0] Channel closed: Failed to read binary packet data!
INFO:asyncssh:[conn=0] Closing connection
INFO:asyncssh:[conn=0] Sending disconnect: Disconnected by application (11)
SFTP operation failed: SSH connection closed
```
this is the code that I ran:
```   
    options = asyncssh.SSHClientConnectionOptions(
        username=username,
        password=password,
        known_hosts=None
    )

 async with asyncssh.connect(sftp_url, port=22, options=options) as conn:
        print("Connected", conn)
        async with conn.start_sftp_client() as sftp:
            # do something
            print("Listdir")
            await sftp.listdir("/sftp")
```
I also tried to specify some arguments that is mentioned https://github.com/ronf/asyncssh/issues/105, but still doesn't work

Meanwhile, when I tried using linux sftp command, I can connect to this sftp server. When I run it using verbose command, seems like everything is supported by asyncssh.

```
debug2: first_kex_follows 0
debug2: reserved 0
debug1: kex: algorithm: ecdh-sha2-nistp256
debug1: kex: host key algorithm: rsa-sha2-512
debug1: kex: server->client cipher: aes256-gcm@openssh.com MAC: <implicit> compression: none
debug1: kex: client->server cipher: aes256-gcm@openssh.com MAC: <implicit> compression: none
debug1: expecting SSH2_MSG_KEX_ECDH_REPLY
```

Note: I have tried to downgrade to 2.2.1 and it's working, but i don't want to do this because I need some fixes that have been fixed from 2.5.0 onwards 

Any ideas on this? Not sure why it is working with 2.2.1 and not with 2.5.0.. and all kex exchanges/host key algos seem to be supported by asyncssh

**Comment by ronf** [comment]
As noted in #105, the "Failed to read binary packet data" is coming from the server, but the previous failure seemed like it was much earlier, while still in the middle of the key exchange. Here, the error seems to be happening later, since you're able to get auth to succeed before it fails. It looks like the problem in this case may be related to requesting the SSH session.

Since you have it working with AsyncSSH 2.2.1, could you capture the AsyncSSH debug log output for both 2.2.1 working and 2.5.0 failing, preferably with a debug level of at least 2? If we can determine what's different in the request made for a new session, there may be a workaround via config options. It might require going to debug level 3 to identify the specific message that the server doesn't like, but we'll see.

If you have access to the server side to capture logs, it would be great if you could do it there as well for both the working and non-working cases.

**Comment by easymavinmind** [comment]
Thanks for the reply @ronf 

As of now, I don't have access to the server side, so all I can do is providing this success log for 2.2.1 with normal logging
```
INFO:asyncssh:Opening SSH connection to mysftp.server, port 22
INFO:asyncssh:[conn=0] Connection to mysftp.server, port 22 succeeded
INFO:asyncssh:[conn=0]   Local address: localhost, port 56498
DEBUG:asyncssh:[conn=0] Requesting key exchange
DEBUG:asyncssh:[conn=0] Received key exchange request
DEBUG:asyncssh:[conn=0] Beginning key exchange
DEBUG:asyncssh:[conn=0] Completed key exchange
INFO:asyncssh:[conn=0] Beginning auth for user User1
DEBUG:asyncssh:[conn=0] Received authentication banner
DEBUG:asyncssh:[conn=0] Trying password auth
INFO:asyncssh:[conn=0] Auth for user User1 succeeded
Connected <asyncssh.connection.SSHClientConnection object at 0x1039b9820>
DEBUG:asyncssh:[conn=0, chan=0] Set write buffer limits: low-water=16384, high-water=65536
INFO:asyncssh:[conn=0, chan=0] Requesting new SSH session
INFO:asyncssh:[conn=0, chan=0]   Subsystem: sftp
INFO:asyncssh.sftp:[conn=0, chan=0] Starting SFTP client
DEBUG:asyncssh.sftp:[conn=0, chan=0] Sending init, version=3
DEBUG:asyncssh.sftp:[conn=0, chan=0] Received version=3, extensions:
DEBUG:asyncssh.sftp:[conn=0, chan=0]   newline@vandyke.com:

Maverick SSHD1.7.29:[conn=0, chan=0]   vendor-id: JADAPTIVE Limited
Connected
DEBUG:asyncssh.sftp:[conn=0, chan=0] Sending opendir for /sftp
INFO:asyncssh:[conn=0, chan=0] Received exit status 0
INFO:asyncssh:[conn=0, chan=0] Received channel close
INFO:asyncssh:[conn=0, chan=0] Channel closed
INFO:asyncssh.sftp:[conn=0, chan=0] SFTP client exited
INFO:asyncssh:[conn=0, chan=0] Closing channel
INFO:asyncssh:[conn=0] Closing connection
INFO:asyncssh:[conn=0] Sending disconnect: Disconnected by application (11)
INFO:asyncssh:[conn=0] Connection closed
SFTP operation failed: /sftp does not exist
```

note: please ignore that the directory doesn't exist

I have tried to investigate the difference between 2.2.1 and 2.5.0 where requesting new SSH session, but still no luck.
Once I have regain the access to the sftp server, I will provide with debug level 3 if needed.

**Comment by ronf** [comment]
Thanks - could you also show the level 2 debug log for the broken case? I'm curious if it is properly requesting the subsystem sftp, and how far it gets after that, if anywhere.

Also, I notice in this case that it is using password auth, where your earlier log showed it using keyboard-interactive auth. Any idea why these would be different? Are you changing any options between the two tests, and are you running them on the same client machine?

Regarding debug level 3, you can do that on the AsyncSSH side even without access to the server. Just call `asyncssh.set_debug_level(3)` after initializing the logging subsystem to level DEBUG.

**Comment by easymavinmind** [comment]
I didn't manage to run asyncssh with level 2 debug, will provide this as well once I regain the access

I have no idea why it was asking for keyboard-interactive auth since I didn't change any options. But looks like it fails during this call

```
SSHClientChannel -> create method -> packet = awaitself._open(b'session')
```

and yes, both tests were running on the same machine.

**Comment by ronf** [comment]
Yeah - that's what it looked like from the original debug log you posted. It seemed like the SFTP server was rejecting the open session request with that "Failed to read binary packet data!" error. So, getting a level 3 debug which shows the raw open request packet could help here in both the working & non-working cases, as well as seeing if there are other packets around there which it might not like.

I've seen one SSH server that was very picky about not allowing SSH_IGNORE messages to be sent in certain places, which AsyncSSH is a bit more aggressive about than other SSH implementations. I don't think that behavior has changed much between 2.2.1 and 2.5.0, but it's possible there's something going on related to that.

There were also some changes in the handling of when AsyncSSH will request a PTY in 2.3.0, but I wouldn't expect that to apply to an SFTP request.

**Comment by easymavinmind** [comment]
@ronf  sorry for the late reply. here are the logs for error case (2.5.0)

```
INFO:asyncssh:[conn=0] Auth for user User1 succeeded
DEBUG:asyncssh:[conn=0] Sending extension info
DEBUG:asyncssh:[conn=0]   global-requests-ok: 
DEBUG:asyncssh:[conn=0, pktid=11] Sent MSG_EXT_INFO (7), 31 bytes
  00000000: 07 00 00 00 01 00 00 00 12 67 6c 6f 62 61 6c 2d  .........global-
  00000010: 72 65 71 75 65 73 74 73 2d 6f 6b 00 00 00 00     requests-ok....
Connected <asyncssh.connection.SSHClientConnection object at 0x7f5628abbe10>
DEBUG:asyncssh:[conn=0, chan=0] Set write buffer limits: low-water=16384, high-water=65536
INFO:asyncssh:[conn=0, chan=0] Requesting new SSH session
DEBUG:asyncssh:[conn=0, chan=0]   Initial recv window 2097152, packet size 32768
DEBUG:asyncssh:[conn=0, pktid=12] Sent MSG_IGNORE (2), 5 bytes
  00000000: 02 00 00 00 00                                   .....
DEBUG:asyncssh:[conn=0, chan=0, pktid=13] Sent MSG_CHANNEL_OPEN (90), 24 bytes
  00000000: 5a 00 00 00 07 73 65 73 73 69 6f 6e 00 00 00 00  Z....session....
  00000010: 00 20 00 00 00 00 80 00                          . ......
DEBUG:asyncssh:[conn=0, pktid=9] Received MSG_DISCONNECT (1), 47 bytes
  00000000: 01 00 00 00 02 00 00 00 22 46 61 69 6c 65 64 20  ........"Failed 
  00000010: 74 6f 20 72 65 61 64 20 62 69 6e 61 72 79 20 70  to read binary p
  00000020: 61 63 6b 65 74 20 64 61 74 61 21 00 00 00 00     acket data!....
DEBUG:asyncssh:[conn=0] Received disconnect: Failed to read binary packet data! (2)
INFO:asyncssh:[conn=0] Connection failure: Failed to read binary packet data!
INFO:asyncssh:[conn=0, chan=0] Closing channel due to connection close
INFO:asyncssh:[conn=0, chan=0] Channel closed: Failed to read binary packet data!
INFO:asyncssh:[conn=0] Closing connection
INFO:asyncssh:[conn=0] Sending disconnect: Disconnected by application (11)
DEBUG:asyncssh:[conn=0, pktid=14] Sent MSG_IGNORE (2), 5 bytes
  00000000: 02 00 00 00 00                                   .....
DEBUG:asyncssh:[conn=0, pktid=15] Sent MSG_DISCONNECT (1), 45 bytes
  00000000: 01 00 00 00 0b 00 00 00 1b 44 69 73 63 6f 6e 6e  .........Disconn
  00000010: 65 63 74 65 64 20 62 79 20 61 70 70 6c 69 63 61  ected by applica
  00000020: 74 69 6f 6e 00 00 00 05 65 6e 2d 55 53           tion....en-US
SFTP operation failed: SSH connection closed

```

and success case (2.2.1)

```
INFO:asyncssh:[conn=0] Auth for user User1 succeeded
Connected <asyncssh.connection.SSHClientConnection object at 0x7ff8d782d8d0>
DEBUG:asyncssh:[conn=0, chan=0] Set write buffer limits: low-water=16384, high-water=65536
INFO:asyncssh:[conn=0, chan=0] Requesting new SSH session
DEBUG:asyncssh:[conn=0, chan=0]   Initial recv window 2097152, packet size 32768
DEBUG:asyncssh:[conn=0, pktid=9] Sent MSG_IGNORE (2), 5 bytes
  00000000: 02 00 00 00 00                                   .....
DEBUG:asyncssh:[conn=0, chan=0, pktid=10] Sent MSG_CHANNEL_OPEN (90), 24 bytes
  00000000: 5a 00 00 00 07 73 65 73 73 69 6f 6e 00 00 00 00  Z....session....
  00000010: 00 20 00 00 00 00 80 00                          . ......
DEBUG:asyncssh:[conn=0, pktid=8] Received MSG_CHANNEL_OPEN_CONFIRMATION (91), 17 bytes
  00000000: 5b 00 00 00 00 00 00 00 00 00 00 00 00 00 00 84  [...............
  00000010: d0                                               .
DEBUG:asyncssh:[conn=0, chan=0]   Initial send window 0, packet size 34000
INFO:asyncssh:[conn=0, chan=0]   Subsystem: sftp
DEBUG:asyncssh:[conn=0, pktid=11] Sent MSG_IGNORE (2), 5 bytes
  00000000: 02 00 00 00 00                                   .....
DEBUG:asyncssh:[conn=0, chan=0, pktid=12] Sent MSG_CHANNEL_REQUEST (98), 27 bytes
  00000000: 62 00 00 00 00 00 00 00 09 73 75 62 73 79 73 74  b........subsyst
  00000010: 65 6d 01 00 00 00 04 73 66 74 70                 em.....sftp
DEBUG:asyncssh:[conn=0, chan=0, pktid=9] Received MSG_CHANNEL_WINDOW_ADJUST (93), 9 bytes
  00000000: 5d 00 00 00 00 00 10 00 00                       ]........
DEBUG:asyncssh:[conn=0, chan=0] Received window adjust of 1048576 bytes, new window 1048576
DEBUG:asyncssh:[conn=0, chan=0, pktid=10] Received MSG_CHANNEL_SUCCESS (99), 5 bytes
  00000000: 63 00 00 00 00                                   c....
INFO:asyncssh.sftp:[conn=0, chan=0] Starting SFTP client
DEBUG:asyncssh.sftp:[conn=0, chan=0] Sending init, version=3
DEBUG:asyncssh:[conn=0, chan=0] Sending 9 data bytes
DEBUG:asyncssh:[conn=0, pktid=13] Sent MSG_IGNORE (2), 5 bytes
  00000000: 02 00 00 00 00                                   .....
DEBUG:asyncssh:[conn=0, chan=0, pktid=14] Sent MSG_CHANNEL_DATA (94), 18 bytes
  00000000: 5e 00 00 00 00 00 00 00 09 00 00 00 05 01 00 00  ^...............
  00000010: 00 03                                            ..
DEBUG:asyncssh.sftp:[conn=0, chan=0] Sent FXP_INIT (1), 5 bytes
  00000000: 01 00 00 00 03                                   .....
DEBUG:asyncssh:[conn=0, chan=0] Reading from channel started
DEBUG:asyncssh:[conn=0, chan=0, pktid=11] Received MSG_CHANNEL_DATA (94), 120 bytes
  00000000: 5e 00 00 00 00 00 00 00 6f 00 00 00 6b 02 00 00  ^.......o...k...
  00000010: 00 03 00 00 00 13 6e 65 77 6c 69 6e 65 40 76 61  ......newline@va
  00000020: 6e 64 79 6b 65 2e 63 6f 6d 00 00 00 02 0d 0a 00  ndyke.com.......
  00000030: 00 00 09 76 65 6e 64 6f 72 2d 69 64 00 00 00 38  ...vendor-id...8
  00000040: 00 00 00 11 4a 41 44 41 50 54 49 56 45 20 4c 69  ....JADAPTIVE Li
  00000050: 6d 69 74 65 64 00 00 00 0d 4d 61 76 65 72 69 63  mited....Maveric
  00000060: 6b 20 53 53 48 44 00 00 00 06 31 2e 37 2e 32 39  k SSHD....1.7.29
  00000070: 00 00 00 00 00 00 00 00                          ........
DEBUG:asyncssh:[conn=0, chan=0] Received 111 data bytes
DEBUG:asyncssh.sftp:[conn=0, chan=0] Received FXP_VERSION (2), 107 bytes
  00000000: 02 00 00 00 03 00 00 00 13 6e 65 77 6c 69 6e 65  .........newline
  00000010: 40 76 61 6e 64 79 6b 65 2e 63 6f 6d 00 00 00 02  @vandyke.com....
  00000020: 0d 0a 00 00 00 09 76 65 6e 64 6f 72 2d 69 64 00  ......vendor-id.
  00000030: 00 00 38 00 00 00 11 4a 41 44 41 50 54 49 56 45  ..8....JADAPTIVE
  00000040: 20 4c 69 6d 69 74 65 64 00 00 00 0d 4d 61 76 65   Limited....Mave
  00000050: 72 69 63 6b 20 53 53 48 44 00 00 00 06 31 2e 37  rick SSHD....1.7
  00000060: 2e 32 39 00 00 00 00 00 00 00 00                 .29........
DEBUG:asyncssh.sftp:[conn=0, chan=0] Received version=3, extensions:

```
hope this can help

**Comment by ronf** [comment]
Thanks!

It looks like the SFTP server in this case (which appears to be SecureFX) might not like the SSH_EXT_INFO message. This extension was added in AsyncSSH release 2.3.0, which is consistent with it working in 2.2.1 but not after that.

This extension info message is only sent when the sender sends an "ext-info-s" in the peer kex algorithms as defined in https://datatracker.ietf.org/doc/html/draft-ssh-ext-info-04, so that should limit it to only being sent to servers that will understand it. Reading the ext-info draft more carefully, though, I see a note in the draft that an SSH_EXT_INFO from the client must be sent immediately after the SSH_NEWKEYS message, so it looks like that may be the problem. AsyncSSH is sending this message after authentication completes. I'll look into fixing this, but I may need your help in testing a proper fix since I don't have access to a SecureFX SFTP server here.

In the meantime, here's a patch which might help you work around the issue. Look for the following code in connection.py in the _process_userauth_success method of the SSHConnection class and remove it:

```python
            if self._can_send_ext_info:
                self._extensions_sent[b'global-requests-ok'] = b''
                self._send_ext_info()
```

**Comment by easymavinmind** [comment]
@ronf  Awesome! It works now.

Thanks for all your help. I hope we can have the next release that includes this fix soon

**Comment by ronf** [comment]
I took a closer look at this today, and there were a couple of problems with the sending of the SSH_EXT_INFO message. In addition to a problem about when it sent SSH_EXT_INFO when acting as a client, it was only sending the server-sig-algs and not the global-requests-ok extension when acting as a server.

Here's an updated patch against the original connection.py (before deleting the lines I suggested you remove above). It still removes those lines, but then adds back sending the SSH_EXT_INFO in the proper places. It also fixes a minor issue with showing that it is sending 'ext-info-c' or 'ext-info-s' in the kex algorithms in the debug output.

Could you apply this patch and then try it against your SFTP server to make sure this still fixes the breakage you saw?

```patch
diff --git a/asyncssh/connection.py b/asyncssh/connection.py
index 2b35b57..19b5cdc 100644
--- a/asyncssh/connection.py
+++ b/asyncssh/connection.py
@@ -1396,6 +1396,7 @@ class SSHConnection(SSHPacketHandler, asyncio.Protocol):
         kex_algs = expand_kex_algs(self._kex_algs, gss_mechs,
                                    bool(self._server_host_key_algs))

+        kex_algs += self._get_ext_info_kex_alg()
         host_key_algs = self._server_host_key_algs or [b'null']

         self.logger.debug1('Requesting key exchange')
@@ -1406,7 +1407,7 @@ class SSHConnection(SSHPacketHandler, asyncio.Protocol):
         self.logger.debug2('  Compression algs: %s', self._cmp_algs)

         cookie = os.urandom(16)
-        kex_algs = NameList(kex_algs + self._get_ext_info_kex_alg())
+        kex_algs = NameList(kex_algs)
         host_key_algs = NameList(host_key_algs)
         enc_algs = NameList(self._enc_algs)
         mac_algs = NameList(self._mac_algs)
@@ -1494,6 +1495,8 @@ class SSHConnection(SSHPacketHandler, asyncio.Protocol):

         self.send_packet(MSG_NEWKEYS)

+        self._extensions_sent[b'global-requests-ok'] = b''
+
         if self.is_client():
             self._send_encryption = next_enc_cs
             self._send_enchdrlen = 1 if etm_cs else 5
@@ -1545,10 +1548,11 @@ class SSHConnection(SSHPacketHandler, asyncio.Protocol):
             if first_kex:
                 self._next_service = _USERAUTH_SERVICE

-                if self._can_send_ext_info:
-                    self._extensions_sent[b'server-sig-algs'] = \
-                        b','.join(self._sig_algs)
-                    self._send_ext_info()
+                self._extensions_sent[b'server-sig-algs'] = \
+                    b','.join(self._sig_algs)
+
+        if self._can_send_ext_info:
+            self._send_ext_info()

         self._kex_complete = True
         self._send_deferred_packets()
@@ -2048,10 +2052,6 @@ class SSHConnection(SSHPacketHandler, asyncio.Protocol):
             self._set_keepalive_timer()
             self._owner.auth_completed()

-            if self._can_send_ext_info:
-                self._extensions_sent[b'global-requests-ok'] = b''
-                self._send_ext_info()
-
             if self._acceptor:
                 result = self._acceptor(self)

```

**Comment by ronf** [comment]
I've gone ahead with this change in commit f91a9f1 in the "develop" branch. If it doesn't work, let me know. Otherwise, I'll plan on including it in the next release.

**Comment by ronf** [comment]
This change is now available in AsyncSSH version 2.7.1.
